### PR TITLE
add initial design overview and threat model

### PR DESCRIPTION
**Pull request by THS-on**
This add new section called design. This includes a more high level description of all the components in Keylime and a basic threat model.

### CODE REVIEW

**Comment by THS-on** [comment]
> So validation the UEFI event log is more secure. You might want to add something about this to the text.

@edwards-n  I hinted to that in the overview, but I agree that it should be more explicit stated in the text.

**Review by aplanas — APPROVED** [review]
There are some minor typos that can be fixed later, but this test is really useful.

**Comment by THS-on** [comment]
@edwards-n I added now a statement that the measured boot policy engine is preferred over static PCR values.

@aplanas If you have the time can you mark the typos? Then I can fix them in this PR.

@kkaarreell I added the inclusion of a nonce in the quote. Is this explanation enough?

**Comment by aplanas** [comment]
>  If you have the time can you mark the typos? 

Sure, I am not English native and is usually others pointing me

File: docs/design/threat_model.rst [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+Keylime was originally developed with the intention of using it in combination with hypervisors to protect the VMs
+against by using the vTPM support in Xen. vTPM support for TPM2.0 was never implemented into Keylime and swtpm+libvirt
+never supported it, so this model no longer fits. Keylime is commonly used either on bare metal hardware or in VMs where
+the TPM is emulated but from VM side treated the same as a hardware TPM. Therefore the common threat model is defined on
+the latter use case.
+
+From Keylime's perspective the core hardware like CPU, memory, motherboard is trusted, because it does not provide
+mechanisms to detect tampering with the hardware itself.
+Keylime chains its root of trust to the TPM therefore the TPM is deemed in general trustworthy. This trust is verified
```
Comment (aplanas) [inline docs/design/threat_model.rst:13]: "into the TPM"?

File: docs/design/threat_model.rst [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+using the EK or EK certificate.
+
+The goal of Keylime is to attest the state of a running system. For this to work the entire boot chain has to be
+verified. The UEFI with Secure Boot enabled firmware and CRTM are generally trusted because it provides the UEFI event
+log and the API for other EFI applications to use the TPM. All the other applications in the boot chain are either
+measured by the firmware or the application that loads them (e.g. GRUB2 loads the kernel). The threat model does not
+require to trust arbitrary EFI applications during because it can be verified after boot what was executed.
```
Comment (aplanas) [inline docs/design/threat_model.rst:20]: "during the boot process"?

File: docs/design/threat_model.rst [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+""""""""""""""""""""""
+Because the TPM is the root-of-trust for Keylime, it ensures that the quote is valid.
+This is vital for all the other attestation steps because the quote is used to validate the data.
+
+Keylime ensures this through three steps:
+
+  * EK validation: The tenant allows Keylime to verify the EK certificate against the CAs of hardware manufacturers
+    or add custom validation steps. This is done to ensure that the EK belongs to an actual hardware TPM or a trusted
+    software TPM.
+  * AK enrollment: Using `MakeCredential`, `ActivateCredential` and enforcing certain key properties
```
Comment (aplanas) [inline docs/design/threat_model.rst:48]: Are "MakeCredential" and "ActivateCredential" TPM2 commands?

**Comment by kkaarreell** [comment]
> @kkaarreell I added the inclusion of a nonce in the quote. Is this explanation enough?

@THS-on yes, thank you.

**Comment by THS-on** [comment]
> @THS-on did you want to fix those typos in this PR or the next?

@mpeters I'll fix them tomorrow in this PR.

File: docs/design/overview.rst [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+getting a uuid for the agent, collecting the EK, EK certificate and AK from an agent
+and verifying that the AK belongs to the provided EK (using `MakeCredential` and `ActivateCredential`).
+
+Once an agent has been registered at the registrar other systems like the verifier and
+tenant can use the EK certificate for checking if they trust that TPM and the AK for
+validating that the signed data sent by an agent is valid.
+
+.. note::
+
+    The Keylime agent currently generates a AK on every startup and sends the EK and
```
[the remark this answers is not in the document]
Comment (THS-on) [reply docs/design/overview.rst:37]: >  IIRC the EK is not sent anywhere? You mean some public part? And is send to where?

Good catch. Yes I mean the public part. I added some clarification.

File: docs/design/overview.rst [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+Measured Boot using the UEFI Event Log
+""""""""""""""""""""""""""""""""""""""
+On larger deployments it is not feasible to collect golden values for the PCR values used for measured boot. To make
+attestation still possible Keylime includes a policy engine for validating the UEFI event log. This is the preferred
+method because static PCR values are fragile and change often if the system changes (firmware, Shim, GRUB, Linux kernel,
+initrd etc.). More information can be found in :doc:`../user_guide/use_measured_boot`.
+
+IMA validation
+""""""""""""""
+Keylime allows to verify files measured by IMA against either an provided allowlist or against a signature.
```
[the remark this answers is not in the document]
Comment (THS-on) [reply docs/design/overview.rst:69]: For signature verification see here: https://github.com/keylime/keylime/pull/718

I think the feature has not made it into the documentation yet.

[the remark this answers is not in the document]
Comment (THS-on) [reply docs/design/overview.rst:69]: The above PR was for automatically learning the keys. The feature for normal keyring feature is only documented here: https://keylime-docs.readthedocs.io/en/latest/rest_apis.html#post--v2.0-agents-agent_id-UUID.

I opened https://github.com/keylime/keylime-docs/issues/79 for adding better documentation on that.

File: docs/design/overview.rst [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+This makes it for example easy to attest all files that were executed by root.
+More information can be found in :doc:`../user_guide/runtime_ima`.
+
+
+Tenant
+------
+The tenant is a commandline management tool shipped by Keylime to manage agents.
+This includes adding and removing the agent from attestation, validation of the EK certificate against a cert store and
```
[the remark this answers is not in the document]
Comment (THS-on) [reply docs/design/overview.rst:77]: In this case it is a cert store.

File: docs/design/threat_model.rst (hunk 3 above)
Comment (THS-on) [reply docs/design/threat_model.rst:48]: Yes

**Pull request merged**